From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Thu, 21 Nov 2024 13:24:47 +0100
Subject: [PATCH 051/225] clk: sprd: add reset support
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add support for resets provided by the clock driver. Imported from
the Spreadtrum downstream kernel.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/clk/sprd/Makefile |   1 +
 drivers/clk/sprd/common.c |   1 +
 drivers/clk/sprd/common.h |   2 +
 drivers/clk/sprd/reset.c  | 105 ++++++++++++++++++++++++++++++++++++++
 drivers/clk/sprd/reset.h  |  30 +++++++++++
 5 files changed, 139 insertions(+)
 create mode 100644 drivers/clk/sprd/reset.c
 create mode 100644 drivers/clk/sprd/reset.h

diff --git a/drivers/clk/sprd/Makefile b/drivers/clk/sprd/Makefile
index f25b2c390..72f6ecf82 100644
--- a/drivers/clk/sprd/Makefile
+++ b/drivers/clk/sprd/Makefile
@@ -7,6 +7,7 @@ clk-sprd-y	+= mux.o
 clk-sprd-y	+= div.o
 clk-sprd-y	+= composite.o
 clk-sprd-y	+= pll.o
+clk-sprd-y	+= reset.o
 
 ## SoC support
 obj-$(CONFIG_SPRD_SC9860_CLK)	+= sc9860-clk.o
diff --git a/drivers/clk/sprd/common.c b/drivers/clk/sprd/common.c
index cd4ed3a7a..ee263305e 100644
--- a/drivers/clk/sprd/common.c
+++ b/drivers/clk/sprd/common.c
@@ -80,6 +80,7 @@ int sprd_clk_regmap_init(struct platform_device *pdev,
 	}
 
 	sprd_clk_set_regmap(desc, regmap);
+	platform_set_drvdata(pdev, regmap);
 
 	data->gate_clk = devm_clk_get_optional_enabled(&pdev->dev, "enable");
 	if (IS_ERR(data->gate_clk))
diff --git a/drivers/clk/sprd/common.h b/drivers/clk/sprd/common.h
index 1f348bdef..260a74eb2 100644
--- a/drivers/clk/sprd/common.h
+++ b/drivers/clk/sprd/common.h
@@ -24,6 +24,8 @@ struct sprd_clk_desc {
 	struct sprd_clk_common		**clk_clks;
 	unsigned long			num_clk_clks;
 	struct clk_hw_onecell_data      *hw_clks;
+	struct sprd_reset_map		*resets;
+	unsigned long			num_resets;
 	bool				need_regcache;
 };
 
diff --git a/drivers/clk/sprd/reset.c b/drivers/clk/sprd/reset.c
new file mode 100644
index 000000000..402b4b472
--- /dev/null
+++ b/drivers/clk/sprd/reset.c
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Spreadtrum reset clock driver
+//
+// Copyright (C) 2022 Spreadtrum, Inc.
+// Author: Zhifeng Tang <zhifeng.tang@unisoc.com>
+
+#include <linux/delay.h>
+#include <linux/regmap.h>
+#include <linux/reset-controller.h>
+
+#include "reset.h"
+
+static inline struct sprd_reset *to_sprd_reset(struct reset_controller_dev *rcdev)
+{
+	return container_of(rcdev, struct sprd_reset, rcdev);
+}
+
+static int sprd_reset_assert(struct reset_controller_dev *rcdev,
+			     unsigned long id)
+{
+	struct sprd_reset *reset = to_sprd_reset(rcdev);
+	const struct sprd_reset_map *map = &reset->reset_map[id];
+	unsigned long flags;
+	unsigned int val = 0;
+
+	spin_lock_irqsave(&reset->lock, flags);
+	regmap_read(reset->regmap, map->reg, &val);
+	val |= map->mask;
+	regmap_write(reset->regmap, map->reg, val);
+	spin_unlock_irqrestore(&reset->lock, flags);
+
+	return 0;
+}
+
+static int sprd_reset_deassert(struct reset_controller_dev *rcdev,
+			       unsigned long id)
+{
+	struct sprd_reset *reset = to_sprd_reset(rcdev);
+	const struct sprd_reset_map *map = &reset->reset_map[id];
+	unsigned long flags;
+	unsigned int val = 0;
+
+	spin_lock_irqsave(&reset->lock, flags);
+	regmap_read(reset->regmap, map->reg, &val);
+	val &= ~map->mask;
+	regmap_write(reset->regmap, map->reg, val);
+	spin_unlock_irqrestore(&reset->lock, flags);
+
+	return 0;
+}
+
+static int sprd_reset_reset(struct reset_controller_dev *rcdev,
+			    unsigned long id)
+{
+	sprd_reset_assert(rcdev, id);
+	udelay(1);
+	sprd_reset_deassert(rcdev, id);
+
+	return 0;
+}
+
+static int sprd_sc_reset_assert(struct reset_controller_dev *rcdev,
+				unsigned long id)
+{
+	struct sprd_reset *reset = to_sprd_reset(rcdev);
+	const struct sprd_reset_map *map = &reset->reset_map[id];
+	unsigned int offset = map->sc_offset;
+
+	return regmap_write(reset->regmap, map->reg + offset, map->mask);
+}
+
+static int sprd_sc_reset_deassert(struct reset_controller_dev *rcdev,
+				  unsigned long id)
+{
+	struct sprd_reset *reset = to_sprd_reset(rcdev);
+	const struct sprd_reset_map *map = &reset->reset_map[id];
+	unsigned int offset = map->sc_offset * 2;
+
+	return regmap_write(reset->regmap, map->reg + offset, map->mask);
+}
+
+static int sprd_sc_reset_reset(struct reset_controller_dev *rcdev,
+			       unsigned long id)
+{
+	sprd_sc_reset_assert(rcdev, id);
+	udelay(1);
+	sprd_sc_reset_deassert(rcdev, id);
+
+	return 0;
+}
+
+const struct reset_control_ops sprd_reset_ops = {
+	.assert		= sprd_reset_assert,
+	.deassert	= sprd_reset_deassert,
+	.reset		= sprd_reset_reset,
+};
+EXPORT_SYMBOL_GPL(sprd_reset_ops);
+
+const struct reset_control_ops sprd_sc_reset_ops = {
+	.assert		= sprd_sc_reset_assert,
+	.deassert	= sprd_sc_reset_deassert,
+	.reset		= sprd_sc_reset_reset,
+};
+EXPORT_SYMBOL_GPL(sprd_sc_reset_ops);
diff --git a/drivers/clk/sprd/reset.h b/drivers/clk/sprd/reset.h
new file mode 100644
index 000000000..cd7083030
--- /dev/null
+++ b/drivers/clk/sprd/reset.h
@@ -0,0 +1,30 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+//
+// Spreadtrum reset clock driver
+//
+// Copyright (C) 2022 Spreadtrum, Inc.
+// Author: Zhifeng Tang <zhifeng.tang@unisoc.com>
+
+#ifndef _SPRD_RESET_H_
+#define _SPRD_RESET_H_
+
+#include <linux/reset-controller.h>
+#include <linux/spinlock.h>
+
+struct sprd_reset_map {
+	u32	reg;
+	u32	mask;
+	u32	sc_offset;
+};
+
+struct sprd_reset {
+	struct reset_controller_dev	rcdev;
+	const struct sprd_reset_map	*reset_map;
+	struct regmap			*regmap;
+	spinlock_t			lock;
+};
+
+extern const struct reset_control_ops sprd_reset_ops;
+extern const struct reset_control_ops sprd_sc_reset_ops;
+
+#endif /* _SPRD_RESET_H_ */
